### PR TITLE
Better support of `optimize_read_in_order` in case of fixed prefix of sorting key

### LINKED ISSUE #38036: optimize_aggregation_in_order for queries with filters
```sql
create table test_agg ( A Int64, B Int64 ) Engine=MergeTree() order by (A, B);
insert into test_agg select intDiv(number, 1e7), number from numbers(1e8);

set optimize_aggregation_in_order=1;

select A, B , count() from test_agg where A = 1 group by A, B limit 1 ;
Elapsed: 0.666 sec. Processed 10.01 million rows,

explain actions=1 select A, B , count() from test_agg where A = 1 group by A, B   limit 1 ;
....
           ReadType: Default


```

if you change predicate to `>=`

```sql
select A, B , count() from test_agg where A>= 1 group by A, B limit 1 ;
Elapsed: 0.069 sec. Processed 1.66 million rows,

explain actions=1 select A, B , count() from test_agg where A>= 1 group by A, B limit 1 ;
...
   ReadType: InOrder
```

`A >= 1 and A <=1`
```sql
select A, B , count() from test_agg where A >= 1 and A <=1 group by A, B   limit 1 ;
Elapsed: 0.063 sec. Processed 745.47 thousand rows

explain actions=1 select A, B , count() from test_agg where A>= 1 group by A, B limit 1 ;
....ReadType: InOrder
```

```sql
select A, B , count() from test_agg where A between 1 and 1 group by A, B   limit 1 ;
Elapsed: 0.064 sec. Processed 745.47 thousand rows
```

```sql
select A, B , count() from test_agg where A in (1)  group by A, B   limit 1 ;
Elapsed: 0.075 sec. Processed 802.82 thousand rows


select A, B , count() from test_agg where A in (1,2)  group by A, B   limit 1 ;
Elapsed: 0.078 sec. Processed 745.47 thousand rows
```

```sql
select  B , count() from test_agg where A =1  group by  B   limit 1 ;
Elapsed: 0.037 sec. Processed 745.47 thousand rows
```

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Better support of `optimize_read_in_order` in case when some of sorting key columns are always constant after applying `WHERE` clause. E.g. query like `SELECT ... FROM table WHERE a = 'x' ORDER BY a, b`, where `table` has storage definition: `MergeTree ORDER BY (a, b)`.

Resolves #38036.

### CODE REVIEW

File: src/Storages/ReadInOrderOptimizer.cpp [hunk 1]
```diff
@@ -180,46 +198,54 @@ InputOrderInfoPtr ReadInOrderOptimizer::getInputOrderImpl(
...
-        int current_direction = matchSortDescriptionAndKey(
-            actions[i]->getActions(), description[i], sorting_key_columns[i + fixed_prefix_size]);
+        int current_direction = matchSortDescriptionAndKey(actions[desc_pos]->getActions(), description[desc_pos], sorting_key_columns[key_pos]);
+        bool is_matched = current_direction && (desc_pos == 0 || current_direction == read_direction);
+
+        if (!is_matched)
+        {
+            /// If one of the sorting columns is constant after filtering,
+            /// skip it, because it won't affect order anymore.
+            if (fixed_sorting_columns.contains(sorting_key_columns[key_pos]))
```
Comment (novikd) [inline src/Storages/ReadInOrderOptimizer.cpp:227]: It seems like we can skip it even if `is_matched` is true, can't we?

Comment (CurtizJ) [reply src/Storages/ReadInOrderOptimizer.cpp:227]: If we skip when `is_matched` is true, `sort_description_for_merging` will probably differ from prefix of global sort description and we will get mismatch of sort descriptions in `FinishSorting` transform. We need somehow to modify global sort description and it's hard to do from this place. I think, it's ok to leave it as is.

**Comment by CurtizJ** [comment]
Integration tests failures: `ERROR: for node1  Get "https://registry-1.docker.io/v2/": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)`.

**Review by novikd — APPROVED** [review]

**Comment by e-mars** [comment]
@novikd @CurtizJ Can this be backported to 22.3 LTS?

**Pull request merged**